pymobiledevice3/remote/tunnel_service.py

```diff
@@ -965,20 +968,38 @@ async def start_tunnel(
         raise Exception(f'Bad value for protocol_handler: {protocol_handler}')
 
 
-async def get_core_device_tunnel_services(bonjour_timeout: float = DEFAULT_BONJOUR_TIMEOUT) \
-        -> List[CoreDeviceTunnelService]:
+async def get_core_device_tunnel_services(bonjour_timeout: float = DEFAULT_BONJOUR_TIMEOUT,
+        udid: Optional[str] = None) -> List[CoreDeviceTunnelService]:
     result = []
-    for rsd in await get_rsds(bonjour_timeout=bonjour_timeout):
-        result.append(create_core_device_tunnel_service_using_rsd(rsd))
+    invalid_service_error = None
+    for rsd in await get_rsds(bonjour_timeout=bonjour_timeout, udid=udid):
+        if udid is None and Version(rsd.product_version) < Version('17.0'):
+            logger.debug(f'Skipping {rsd.udid}:, iOS {rsd.product_version} < 17.0')
+            rsd.close()
+            continue
+        try:
+            result.append(create_core_device_tunnel_service_using_rsd(rsd))
+        except InvalidServiceError as e:
+            invalid_service_error = invalid_service_error or e
+            logger.error(f'Skipping {rsd.udid} iOS {rsd.product_version}: {e}')
+            rsd.close()
+        except Exception as e:
+            logger.error(f'Failed to start service: {rsd}')
+            raise
+    # Only raise if all attached devices fail. May lead to retrying with --tunnel argument.
+    if invalid_service_error and not result:
```

Comment:
this will raise error only if now devices were not making this function behavior not deterministic. let's instead just report the errors as debug-level since they are considered legit for some of the devices and return the list as we get it. no errors should be raised

Reply:
This mixes with my other comment.  If >= iOS 17, they are always errors(InvalidServiceError should never be thrown).  On older iOS, they are "debug" because we expect those devices will probably fail... (which is why I think they should be skipped by default).

There is a bigger side effect to what you are suggesting.  If we never raise InvalidServiceError, then it won't be caught in main(), and you won't print the diagnostic INVALID_SERVICE_MESSAGE at all.  You'll just get back the generic `ERROR Device is not connected` that is printed when no matching devices are found.  That's why I took the extra steps to throw it.

The user really needs to see some output for every iOS17 device, especially if there is an error.  If they have 3 iOS17 devices plugged in, and run `start-tunnel`, and only two devices appear on the list, they will wonder what happened to the 3rd.  That's why I chose .error() over .debug().

Reply:
From my understanding, since we now agree on that only iOS 17 should be discovered, then all errors should be thrown always instead of ignoring some of them. All iOS 17 devices should succeed in establishing the tunnel

Reply:
Yes, I agree.  Once I added the iOS17 check, I thought about just skipping the InvalidServiceError check entirely.  But you and I had already talked about it catching those errors, so I thought I'd try anyway.

I'll remove `except InvalidServiceError` and submit an updated patch.